File: cliente.js
const net = require('net');
const readline = require('readline');

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout
});

function ask(question) {
  return new Promise(resolve => rl.question(question, resolve));
}

async function main() {
  const host = (await ask('IP del servidor [127.0.0.1]: ')).trim() || '127.0.0.1';
  const port = parseInt((await ask('Puerto del servidor [8000]: ')).trim(), 10) || 8000;
  const nick = (await ask('Ingresa tu nombre: ')).trim() || 'Jugador';
  rl.close();

  const client = new net.Socket();

  client.on('data', (data) => {
    const msg = data.toString().trim();
    if (msg) {
      process.stdout.write('\r' + msg + '\n');
      process.stdout.write('Cliente > ');
    }
  });

  client.on('error', (err) => {
    console.error('Error de conexión:', err.message);
    process.exit(1);
  });

  client.on('close', () => {
    console.log('\nConexión cerrada.');
    process.exit(0);
  });

  client.connect(port, host, () => {
    client.write(nick + '\n');
    console.log(`Conectado a ${host}:${port}`);
    console.log('Escribe un mensaje y presiona Enter para enviar.\n');

    const chat = readline.createInterface({
      input: process.stdin,
      output: process.stdout,
      prompt: 'Cliente > '
    });

    chat.prompt();

    chat.on('line', (line) => {
      if (line.trim() === '') {
        chat.prompt();
        return;
      }
      client.write(nick + ': ' + line + '\n');
      chat.prompt();
    });

    chat.on('close', () => {
      client.end();
      process.exit(0);
    });
  });
}

main();

File: cliente.go
package main

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"strings"
)

func ask(prompt, def string) string {
	fmt.Printf("%s", prompt)
	line, _ := bufio.NewReader(os.Stdin).ReadString('\n')
	line = strings.TrimSpace(line)
	if line == "" {
		return def
	}
	return line
}

func main() {
	host := ask("IP del servidor [127.0.0.1]: ", "127.0.0.1")
	port := ask("Puerto del servidor [8000]: ", "8000")
	nick := ask("Ingresa tu nombre: ", "Jugador")

	conn, err := net.Dial("tcp", net.JoinHostPort(host, port))
	if err != nil {
		fmt.Println("Error de conexión:", err)
		os.Exit(1)
	}
	defer conn.Close()
	conn.Write([]byte(nick + "\n"))

	fmt.Printf("Conectado a %s:%s\n", host, port)
	fmt.Println("Escribe un mensaje y presiona Enter para enviar.")

	go func() {
		reader := bufio.NewReader(conn)
		for {
			msg, err := reader.ReadString('\n')
			if err != nil {
				fmt.Println("\nConexión cerrada.")
				os.Exit(0)
			}
			fmt.Printf("\r%sCliente > ", msg)
		}
	}()

	scanner := bufio.NewScanner(os.Stdin)
	for {
		fmt.Print("Cliente > ")
		if !scanner.Scan() {
			break
		}
		line := strings.TrimRight(scanner.Text(), "\n")
		if strings.TrimSpace(line) == "" {
			continue
		}
		conn.Write([]byte(nick + ": " + line + "\n"))
	}
}
